### ISSUE TITLE
Windows can't use the new lfe and lfec scripts

### ISSUE THREAD

**Issue by oubiwann** [lfe/lfe#138]
Due to Windows lacking support for a complete POSIX `exec`, running `lfe` actually only starts the Erlang shell. Instead a `.bat` file needs to be created (for both `lfe` and `lfec`).

I will start work on this and track progress here. I will be using RabbitMQ's `.bat` as a source of inspiration: 
- https://github.com/rabbitmq/rabbitmq-server/blob/master/scripts/rabbitmq-server.bat

**Comment by rvirding** [comment]
Yes, this we really need. I suppose the only question whether we do everything in .bat files equivalent to lfe or lfec, or should the .bat files just call (a working) lfeexec. If we did the latter then I would fix it for both unixes and windows and use it from lfe/lfec.

**Comment by oubiwann** [comment]
Yeah, `lfeexec` does compile on Windows, and if that was working, that might be a nice general solution ... (it causes an Erlang crash dump on Windows right now).

For now, I'll go with a minimally working `.bat` but I'll be at the ready for converting to `lfeexec`.

**Comment by yurrriq** [comment]
Note to self: update xlfe instructions accordingly once this is resolved.

**Comment by oubiwann** [comment]
This is a bit more complicated that anticipated; I've asked for some help/insights from both 1) Windows users, and 2) those who know more about the `-user` flag for `erl`:
- https://groups.google.com/d/msg/lisp-flavoured-erlang/FuhQYCK-Svo/o9wH5hnFCAAJ

**Comment by rvirding** [comment]
About windows I can't help you but with `-user` I can. `-user` is used to specify which top-level user interface to start. If you don't give the flag then the default with the ^G and multishell version is used.

`lfe` does a `-user lfe_init` which means that `lfe_int:start/0` will be called. `lfe_init:start/0` first checks if we want to evaluate a string, given with `-lfe_eval`, or run a script. When running a script you don't want the default terminal interface but a simpler one which is started with `user:start()`. If neither then it calls the default top-level terminal interface, the one with ^G handling, but with the LFE shell instead of the normal erlang shell. The `'tty_sl -c -e'` means start up a character based terminal input, again this is the same as the default.

The whole startup mechanism is a bit messy because there are no clear orthogonal but a mishmash of overlapping options with a lot of "extra" functionality.

My guess is that somehow the windows `erl` doesn't "see" the `-user` option so it goes into the default handling with the erlang shell. We clearly need a proper .bat file for this or we fix up `lfeexec`. I can do the lfeexec if necessary but .bat help is definitely needed. First step what do the startup files look like under windows? Maybe we can copy them. Unfortunately I don't have access to a windows I can play with but I will try and fix this.

**Comment by yurrriq** [comment]
~~I've got a Windows 10 box and access to a Win8 VM and can help test.~~ Dead until further notice.

**Comment by ivanhernandez** [comment]
`lfeexec.exe` does compile now and starts ok, for me (warning about a missing `LFE_ROOT`).

Setting this environment variable to be the absolute path where parent of `ebin` folder is, `lfeexec` starts EShell, but LFE shell is not loaded.

I have tried both with Windows and UNIX style slashes. As you explained, `-user lfe_init` option is not taken into consideration.

Do not know a thing about all of this, but I would perform tests in my Windows box, if you wanted to. I compiled lteexec with the latest TDM-GCC for x64.

**Comment by rvirding** [comment]
A quick comment about this. The LFE_BOOT environment variable is actually not used by LFE, it assumes all the relevant directories are in the code load path either by explicitly adding them with -på or adding them to the ERL_LIBS environment variable.  Check the lfe she'll script.

**Comment by gargakshit** [comment]
I made a fork with some windows specific fixes: https://github.com/gargakshit/lfe. Please have a look at the commit message on how it works. You need to use native bash (git-bash, msys, etc) on windows to use it. I want to PR it, but I need maintainer approval and need someone to review it first.

PS: It feels weird to comment after 6 years

**Comment by rvirding** [comment]
@gargakshit Unfortunately my Windows knowledge isn't enough to test this solution if someone else could I wouldn't have anything against merging it.
